File: FE/DATN-Fashion-Shop-FE/src/app/models/Category/Category.ts
import {CategoryTranslation} from './CategoriesTranslation';

export interface Category {
  id: number;
  imageUrl: string;
  isActive: boolean;
  parentCategory: Category | null;
  translations: CategoryTranslation[];
  subCategories: Category[];

  getTranslationByLanguage(languageCode: string): string | null;
  getAllSubCategories(): Category[];
}

export class CategoryImpl implements Category {
  id: number;
  imageUrl: string;
  isActive: boolean;
  parentCategory: Category | null;
  translations: CategoryTranslation[];
  subCategories: Category[];

  constructor(data?: Partial<Category>) {
    this.id = data?.id || 0;
    this.imageUrl = data?.imageUrl || '';
    this.isActive = data?.isActive ?? true;
    this.parentCategory = data?.parentCategory ? new CategoryImpl(data.parentCategory) : null;
    this.translations = data?.translations || [];
    this.subCategories = data?.subCategories?.map(sub => new CategoryImpl(sub)) || [];
  }

  getTranslationByLanguage(languageCode: string): string | null {
    const translation = this.translations.find(
      (t) => t.language.code === languageCode
    );
    return translation ? translation.name : null;
  }

  getAllSubCategories(): Category[] {
    const subCategories: Category[] = [];
    const collectSubCategories = (category: Category) => {
      subCategories.push(category);
      category.subCategories.forEach(collectSubCategories);
    };
    this.subCategories.forEach(collectSubCategories);
    return subCategories;
  }
}

File: DATN-Fashion-Shop-BE/src/main/java/com/example/DATN_Fashion_Shop_BE/model/Category.java
package com.example.DATN_Fashion_Shop_BE.model;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.NotFound;
import org.hibernate.annotations.NotFoundAction;
import org.hibernate.envers.Audited;
import org.hibernate.envers.NotAudited;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@Entity
@Table(name = "categories")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Audited
public class Category extends BaseEntity{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "image_url", length = 255)
    private String imageUrl;

    @Audited
    @Column(name= "is_active")
    private Boolean isActive = true;

    @ManyToOne
    @JoinColumn(name = "parent_id")
    private Category parentCategory;

    @ManyToMany(mappedBy = "categories")
    private Set<Product> products = new HashSet<>();

    @OneToMany(mappedBy = "category", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    private List<CategoriesTranslation> translations;

    @Version
    private Long version;

    public String getTranslationByLanguage(String languageCode) {
        return translations.stream()
                .filter(t -> t.getLanguage().getCode().equals(languageCode))
                .map(CategoriesTranslation::getName)
                .findFirst()
                .orElse(null); // Hoặc giá trị mặc định
    }

    public Set<Category> getAllSubCategories() {
        Set<Category> subCategories = new HashSet<>();
        for (Category child : this.getSubCategories()) {
            subCategories.add(child);
            subCategories.addAll(child.getAllSubCategories()); // Recursively adding subcategories
        }
        return subCategories;
    }

    @OneToMany(mappedBy = "parentCategory")
    private Set<Category> subCategories = new HashSet<>();
}
